### ISSUE TITLE
[FEATURE] @EnumNameConstants

### ISSUE THREAD

**Issue by daniel-shuy** [rzwitserloot/lombok#2731]
**Describe the feature**
Currently, there are 2 main ways to define String constants in Java:
- enums
- static final constants

Enums can be enumerated (using `Enum#values()`, `EnumSet`, etc), and provides better typesafety (enum method parameters, exhaustive switch case with Java 13 Enhanced Switch), but takes more memory (which is usually not an issue except in Android or embedded environments), and its main issue is that it cannot be used as a String constant in annotation parameters (because `Enum#name()` is not a compile time constant).

WIth static final constants, for Android, some of the typesafety can be supplemented with Android's TypeDef annotations, but for non-Android environments, there is no alternative. Manually definiting a Collection to enumerate static final constants is also error prone. A `HashSet`/`HashMap` is also slower than an `EnumSet`/`EnumMap`. 

I suggest to create an annotation called `@EnumNameConstants` (inspired by [@FieldNameConstants](https://projectlombok.org/features/experimental/FieldNameConstants)), that when annotated on an enum, will generate an inner type with string constants of the name of each enum value.
eg.
```java
@EnumNameConstants
public enum Level {
    LOW,
    MEDIUM,
    HIGH,
    ;
}
```
should generate
```java
public enum Level {
    LOW,
    MEDIUM,
    HIGH,
    ;

    public static final class Fields {
        public static final String LOW = "LOW";
        public static final String MEDIUM = "MEDIUM";
        public static final String HIGH = "HIGH";
    }
}
```

This will provide all the advantages of using enums, while removing its greatest limitation.

**Describe the target audience**
Example use case:
```java
public enum Permissions {
    READ_USERS,
    WRITE_USERS,
    ;
}

@Test
public void testUserService() {
    val userService = new UserService();
    userService.createUserWithPermissions(Permissions.values());
}

@Test
@WithMockUser(roles = Permissions.Fields.READ_USERS)
public void test() {
    // ...
}
```

**Additional context**

**Comment by Rawi01** [comment]
You can use `@FieldNameConstants` for this:
```java
@FieldNameConstants
public enum Permissions {
    @FieldNameConstants.Include READ_USERS,
    @FieldNameConstants.Include WRITE_USERS,
    ;
}
```
I don't know if this is intended, I was acutally supprised that it seems to be okay to add annotations to enum values.

---
In general I think that it would be way better if you could define your own annotation that accepts an enum attribute (`@YourWithMockUser(roles = Permissions.READ_USERS)`) and lombok maps that one to the string based annotation (`@WithMockUser(roles = "READ_USERS")`). This might be an extension to the already quite complex meta-annotation feature request.

**Comment by daniel-shuy** [comment]
@Rawi01 I just tried that with the latest Lombok version (`1.18.18`) and it doesn't work for me (I can annotate an enum value with `@FieldNameConstants.Include`, but it won't generate a constant for it). What version did you try with?

**Comment by pkordis-r7** [comment]
> In general I think that it would be way better if you could define your own annotation that accepts an enum attribute...

I don't think that javac allows `enum`s as annotation attributes, in fact it will only allow constants

**Comment by ravidesai47** [comment]
This is one of the important feature for our projects for maintainability. As for all enums we do create which is requested by requestor. I would request lombok team to build this feature.

**Comment by Rawi01** [comment]
@daniel-shuy I just tried it again using 1.18.20 and it works in Eclipse and in `javac` it but it is marked as error in IntelliJ. It still works if you simply run it. Do you think it would be sufficient if we simply extend `@FieldNameConstants` to also pick up the `enum` values?
@pkordis-r7 You can use enums in annotations: [JLS](https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.6.1)
@ravidesai47 Can you share an example why you need this feature? So far there only is `@WithMockUser` and some usage examples would be nice and _might_ increase the prority of this feature request :smile:

**Comment by daniel-shuy** [comment]
@Rawi01
> I just tried it again using 1.18.20 and it works in Eclipse and in javac it but it is marked as error in IntelliJ. It still works if you simply run it.

Interesting, looks like its an issue with the `lombok-intellij-plugin`. But then again, as you previously mentioned, it seems to be an unintended feature.

> Do you think it would be sufficient if we simply extend @FieldNameConstants to also pick up the enum values?

Sure, whatever works. Maybe `@FieldNameConstants` can be extended with a flag to include static fields (which includes enum values)?

> Can you share an example why you need this feature? So far there only is @WithMockUser and some usage examples would be nice and might increase the prority of this feature request 😄

I don't know about @ravidesai47's use case, but I can add one more if it might help increase the priority of this request 😝 
Another example is when using Jackson's polymorphic serialization/deserialization with logical type names (`JsonTypeInfo.Id.NAME`), eg.
```java
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME)
@JsonSubTypes( {
        @JsonSubTypes.Type(value = Automobile.class),
        @JsonSubTypes.Type(value = Plane.class),
        @JsonSubTypes.Type(value = Ship.class),
})
@Data
abstract class Vehicle {
    private Double latitude;
    private Double longitude;
}

enum VehicleType {
    AUTOMOBILE,
    PLANE,
    SHIP,
    ;
}

@JsonTypeName(VehicleType.AUTOMOBILE)
class Automobile {
    private Integer wheelCount;
}
```

This will cause the `Automobile` class to be serialized with a `"@type": "AUTOMOBILE"` field, which can then be used by a typed client (eg. Java, TypeScript, etc) to correctly deserialize it to the right class.

Why is there a need for an enum? There is currently no way to get all subclasses of `Vehicle`. It may eventually be possible if Java's sealed classes (added in Java 15) copies Kotlin's [KClass#sealedSubclasses](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect/-k-class/sealed-subclasses.html), but I'm not holding my breath (this may be another feature that Lombok could implement). `Enum#values()` provides us a way to get all `VehicleType`s.

**Comment by ravidesai47** [comment]
My use-case is related to jackson json polymorphism. We have lot of json polymorphed objects which has a polymorphism based on Enum which requires this feature. We create enum name text fields in all such enums to use them in @JsonSubTypes annotation.

**Comment by daniel-shuy** [comment]
@ravidesai47 oh wow, I'm surprised we have the same use case 😄

**Comment by ravidesai47** [comment]
@daniel-shuy I think it needs to be discussed in https://groups.google.com/g/project-lombok forum. Someone from team project lombok would be able to help there increasing priority of this feature.

**Comment by alexandrenavarro** [comment]
+1 I have exactly the same use case with jackson json polymorphism (and some on others annotations).

**Comment by neseleznev** [comment]
+1 This feature is needed to reference particular enum value in OpenApi descriptions as they're also annotations-based Strings with same limitations
```java
    @Schema(title = "Your enum property",
            description = """
                Yada-yada. \s
                The property appeared on Jan 1, 2023. \s
                For the legacy entities, special value \s
            """ + YourEnumType.LEGACY.name() + " is returned",
            required = true,
            example = YourEnumType.USEFUL_EXAMPLE_VALUE.name()) // NB! also useful!
    @NotNull
    YourEnumType yourEnumType;
```

**Comment by delyand** [comment]
+1 we have a very similar use case. Does this feature have a shot?

**Comment by xerZV** [comment]
+1

**Comment by LuisEscamillaMartin** [comment]
+1

**Comment by hitakashio** [comment]
The solution, @Rawi01 suggested can be used in IntelliJ if you explicitly set onlyExplicitlyIncluded to true:
```java
    @FieldNameConstants(onlyExplicitlyIncluded = true)
    public enum Role{
        @FieldNameConstants.Include ADMIN,
        @FieldNameConstants.Include USER
    }

    private String foo(){
        return Role.Fields.USER;
    }
```

**Comment by daniel-shuy** [comment]
> The solution, @Rawi01 suggested can be used in IntelliJ if you set `onlyExplicitlyIncluded=true`:
> 
> ```java
> @FieldNameConstants(onlyExplicitlyIncluded = true)
> public enum Role{
>         @FieldNameConstants.Include ADMIN,
>         @FieldNameConstants.Include USER
> }
> 
> private String foo(){
>         return Role.Fields.USER;
> }
> ```

It works! 🤯

**Comment by muhrifqii** [comment]
I believe `FieldNameConstants` won't work if we want the variable as constant

```java
@FieldNameConstants(onlyExplicitlyIncluded = true)
@AllArgsConstructor
@Getter
public enum Role{
        @FieldNameConstants.Include ADMIN("role_admin"),
        @FieldNameConstants.Include USER("role_user");
        private final String role;
}

private String foo(){
        return Role.Fields.USER;
}
```
This foo will resulting the enum name "USER" but we want it to be "role_user"
